File: app/src/main/java/com/yxm/customview/algorithm/algorithms/getoffer/GetOffer.java
package com.yxm.customview.algorithm.algorithms.getoffer;

import com.yxm.customview.algorithm.datastruct.binarytree.BST;

import java.util.*;

/**
 * Created by Myron at 2021/3/5 20:10
 *
 * @email yxmbest@163.com
 * @description 剑指offer 题集
 */


public class GetOffer {

    /**
     * 01 二维数组中的查找
     * 在一个二维数组中（每个一维数组的长度相同），
     * 每一行都按照从左到右递增的顺序排序，
     * 每一列都按照从上到下递增的顺序排序。
     * 请完成一个函数，输入这样的一个二维数组和一个整数，判断数组中是否含有该整数。
     */
    public static boolean find(int target, int[][] array) {
        int rows = array.length;
        if (rows == 0)
            return false;
        int cols = array[0].length;
        if (cols == 0)
            return false;

        int row = array.length - 1;
        int column = 0;
        while (row >= 0 && column < cols) {
            if (array[row][column] == target) {
                return true;
            } else if (array[row][column] > target) {
                row--;
            } else {
                column++;
            }
        }

        return false;
    }

    /**
     * 输入一个递增排序的数组和一个数字S，在数组中查找两个数，使得他们的和正好是S，
     * 如果有多对数字的和等于S，输出两个数的乘积最小的。
     * 双指针
     */
    public static ArrayList<Integer> FindNumbersWithSum(int[] array, int sum) {
        ArrayList<Integer> list = new ArrayList<Integer>();
        if (array.length == 0) {
            return list;
        }
        int left = 0;
        int right = array.length - 1;
        while (left < array.length && right >= 0) {
            if (array[left] + array[right] == sum) {
                list.add(array[left]);
                list.add(array[right]);
                break;
            } else if (array[left] + array[right] > sum) {
                right--;
            } else {
                left++;
            }
        }
        return list;
    }


    public static String ReverseSentence(String str) {
        if (str.length() <= 0 || str.trim().equals("")) {
            return str;
        }
        StringBuilder result = new StringBuilder();
        Stack<String> stack = new Stack<>();
        String[] array = str.split(" ");
        for (String s : array) {
            stack.push(s);
        }
        while (!stack.isEmpty()) {
            result.append(stack.pop()).append(" ");
        }
        return result.toString().substring(0, result.length() - 1);
    }

    public static void main(String[] args) {
        int[][] array = new int[3][3];

        for (int i = 0; i < array.length; i++) {
            for (int j = 0; j < array.length; j++) {
                array[i][j] = 2;
            }
        }

        boolean find = find(2, array);
        System.out.println(find);
    }
}

File: app/src/main/java/com/yxm/customview/algorithm/algorithms/getoffer/GetOffer.kt
package com.yxm.customview.algorithm.algorithms.getoffer

import java.util.*

/**
 * Created by Myron at 2021/3/5 20:10
 * @email yxmbest@163.com
 * @description 剑指offer 题集
 */

/**
 * 01 二维数组中的查找
 * 在一个二维数组中（每个一维数组的长度相同），
 * 每一行都按照从左到右递增的顺序排序，
 * 每一列都按照从上到下递增的顺序排序。
 * 请完成一个函数，输入这样的一个二维数组和一个整数，判断数组中是否含有该整数。
 *
 * 二分查找思路，从左下角开始
 */
fun find(target: Int, array: Array<IntArray>): Boolean {
    // write code here
    val length = array.size
    if (length == 0) {
        return false
    }
    val cols = array[0].size
    if (cols == 0) {
        return false
    }
    var column = 0
    var row = array.size - 1
    while (row >= 0 && column < array[0].size) {
        when {
            array[row][column] == target -> {
                return true
            }
            array[row][column] > target -> {
                row--
            }
            else -> {
                column++
            }
        }
    }
    return false
}

/**
 * 请实现一个函数，将一个字符串中的每个空格替换成“%20”。
 * 例如，当字符串为We Are Happy.则经过替换之后的字符串为We%20Are%20Happy。
 * @param s String
 * @return String
 */
fun replaceSpace(s: String): String {
    // write code here
    val sb = StringBuilder()
    for (c in s) {
        if (c == ' ') {
            sb.append("%20")
        }else{
            sb.append(c)
        }
    }
    return sb.toString()
}

/**
 * 牛客最近来了一个新员工Fish，每天早晨总是会拿着一本英文杂志，写些句子在本子上。
 * 同事Cat对Fish写的内容颇感兴趣，有一天他向Fish借来翻看，但却读不懂它的意思。
 * 例如，“student. a am I”。后来才意识到，这家伙原来把句子单词的顺序翻转了，
 * 正确的句子应该是“I am a student.”。Cat对一一的翻转这些单词顺序可不在行，
 * 你能帮助他么？
 * @param str String
 * @return String?
 *
 * 还可用滑动窗口,遇到“ ”就反转字符串
 */
fun ReverseSentence(str: String): String? {
    if (str.isEmpty() || str.trim { it <= ' ' } == "") {
        return str
    }
    val result = java.lang.StringBuilder()
    val stack = Stack<String>()
    val array = str.split(" ".toRegex()).toTypedArray()
    for (s in array) {
        stack.push(s)
    }
    while (!stack.isEmpty()) {
        result.append(stack.pop()).append(" ")
    }
    return result.toString().substring(0, result.length - 1)
}


fun main() {
    print(ReverseSentence("student. a am I"))
}